engine/crates/engine/src/resolver_utils/joins.rs

```diff
@@ -0,0 +1,165 @@
+use std::{
+    collections::BTreeMap,
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+use engine_parser::{
+    types::{Field, Selection, SelectionSet},
+    Positioned,
+};
+use engine_value::{argument_set::ArgumentSet, ConstValue, Name, Value};
+
+use crate::{
+    registry::resolvers::{join::JoinResolver, ResolvedValue},
+    resolver_utils::field::run_field_resolver,
+    Context, ContextField, Error,
+};
+
+#[async_recursion::async_recursion]
+pub async fn resolve_joined_field(
+    ctx: &ContextField<'_>,
+    join: &JoinResolver,
+    parent_resolve_value_for_join: ResolvedValue,
+) -> Result<ResolvedValue, Error> {
+    let mut query_field = fake_query_ast(ctx.item, join, parent_resolve_value_for_join)?;
+    let mut field_iter = join.fields.iter().peekable();
+    let mut resolved_value = ResolvedValue::default();
+    let mut current_type = ctx
+        .schema_env()
+        .registry
+        .root_type(engine_parser::types::OperationType::Query);
+
+    while let Some((name, _)) = field_iter.next() {
+        let meta_field = current_type
+            .field(name)
+            .ok_or_else(|| Error::new(format!("Internal error: could not find joined field {}", &name)))?;
```

Comment:
nitpick, but it would be nice to have the parent type name in the error (`<current_type>.<field_name>`)

Reply:
Yeah, true.   This error shouldn't _really_ happen but I'll improve the error anyway (wouldn't be the first time something that shouldn't happen does actually happen in the end)